server/src/main/java/com/thoughtworks/go/server/persistence/MaterialRepository.java

```diff
@@ -1025,4 +1025,19 @@ public File folderFor(Material material) {
         MaterialInstance materialInstance = this.findOrCreateFrom(material);
         return new File(new File("pipelines", "flyweight"), materialInstance.getFlyweightName());
     }
+
+    public List<Modification> getModificationWithMaterial() {
+        String queryString = "SELECT mods.* " +
+                "FROM (" +
+                "   SELECT MAX(id) OVER (PARTITION BY materialid) as max_id, * " +
+                "   FROM modifications " +
+                ") mods " +
+                "JOIN materials m ON mods.materialid=m.id " +
+                "WHERE mods.id=mods.max_id;";
```

Comment:
We should run this query on larger database and check the performance, also run `EXPLAIN ANALYZE` on PostgreSQL

Reply:
H2: Took the latest backup of build.gocd and ran the query currently used in MaterialRepository
No. of materials: 1411
No. of modifications: 174123
Query metrics: **1145 rows, 1539 ms**


For postgres - had a backup with :
No. of materials: 2696
No. of modifications: 1298467
Ran `EXPLAIN ANALYZE`:
```text
Merge Join  (cost=0.71..186198.21 rows=6490 width=1118) (actual time=23.753..2691.740 rows=2573 loops=1)
   Merge Cond: (m.id = mods.materialid)
   ->  Index Only Scan using newmaterials_pkey on materials m  (cost=0.28..183.66 rows=2696 width=8) (actual time=0.978..2.985 rows=2696 loops=1)
         Heap Fetches: 2696
   ->  Materialize  (cost=0.43..185926.90 rows=6490 width=1118) (actual time=22.769..2686.935 rows=2574 loops=1)
         ->  Subquery Scan on mods  (cost=0.43..185910.67 rows=6490 width=1118) (actual time=22.761..2686.148 rows=2574 loops=1)
               Filter: (mods.id = mods.max_id)
               Rows Removed by Filter: 1295893
               ->  WindowAgg  (cost=0.43..169684.81 rows=1298069 width=1118) (actual time=19.412..2600.758 rows=1298467 loops=1)
                     ->  Index Scan using idx_mod_new_material_id on modifications  (cost=0.43..150213.77 rows=1298069 width=1110) (actual time=0.062..1377.801 rows=1298467 loops=1)
 Planning time: 6.731 ms
 Execution time: 2692.038 ms
(12 rows)
```

Reply:
We will have to look at improving this query.